src/tree/branch.rs

```diff
@@ -0,0 +1,145 @@
+use canonical::{Canon, Store};
+use core::borrow::Borrow;
+use core::ops::Deref;
+use dusk_plonk::prelude::BlsScalar;
+use hades252::{ScalarStrategy, Strategy};
+use microkelvin::{Annotation, Branch};
+use nstack::NStack;
+
+/// Represents a level of a branch on a given depth
+#[derive(Debug, Default, Clone, Copy)]
+pub struct PoseidonLevel {
+    level: [BlsScalar; hades252::WIDTH],
+    offset: usize,
+}
+
+impl PoseidonLevel {
+    /// Represents the offset of a node for a given path produced by a branch
+    /// in a merkle opening
+    pub fn offset(&self) -> usize {
+        self.offset
+    }
+}
+
+impl Deref for PoseidonLevel {
+    type Target = BlsScalar;
+
+    fn deref(&self) -> &Self::Target {
+        &self.level[self.offset]
+    }
+}
+
+impl AsRef<[BlsScalar]> for PoseidonLevel {
+    fn as_ref(&self) -> &[BlsScalar] {
+        &self.level
+    }
+}
+
+/// Represents a full path for a merkle opening
+#[derive(Debug, Clone, Copy)]
+pub struct PoseidonBranch<const DEPTH: usize>([PoseidonLevel; DEPTH]);
+
+impl<const DEPTH: usize> PoseidonBranch<DEPTH> {
+    /// Represents the root for a given path of an opening over a subtree
+    pub fn root(&self) -> BlsScalar {
+        *self.0[DEPTH - 1]
+    }
+}
+
+impl<const DEPTH: usize> Deref for PoseidonBranch<DEPTH> {
+    type Target = BlsScalar;
+
+    fn deref(&self) -> &Self::Target {
+        self.0[0].deref()
+    }
+}
+
+impl<const DEPTH: usize> Default for PoseidonBranch<DEPTH> {
+    fn default() -> Self {
+        PoseidonBranch([PoseidonLevel::default(); DEPTH])
+    }
+}
+
+impl<const DEPTH: usize> AsRef<[PoseidonLevel]> for PoseidonBranch<DEPTH> {
+    fn as_ref(&self) -> &[PoseidonLevel] {
+        &self.0
+    }
+}
+
+impl<L, A, S, const DEPTH: usize> From<&Branch<'_, NStack<L, A, S>, S, DEPTH>>
+    for PoseidonBranch<DEPTH>
+where
+    L: Canon<S>,
+    L: Clone,
+    for<'a> &'a L: Into<BlsScalar>,
+    A: Canon<S>,
+    A: Annotation<NStack<L, A, S>, S>,
+    A: Borrow<BlsScalar>,
+    S: Store,
+{
+    fn from(b: &Branch<'_, NStack<L, A, S>, S, DEPTH>) -> Self {
+        let mut branch = PoseidonBranch::default();
+        let mut depth = 0;
+
+        b.levels()
+            .iter()
+            .rev()
+            .zip(branch.0.iter_mut())
+            .for_each(|(l, b)| {
+                depth += 1;
+                b.offset = l.offset() + 1;
+
+                let mut flag = 1;
+                let mut mask = 0;
+
+                match &**l {
```

Comment:
This looks a bit odd. Wouldn't it be more readable if we call as_ref or deref here?

Reply:
I agree this style is weird and ugly, but as discussed we don't really have a better option for rust